Add Serial API post (#3688)
GoogleChrome/web.dev: src/site/content/en/blog/serial/index.md

To send data to a serial device, pass data to
`port.writable.getWriter().write()`. Calling `releaseLock()` on
`port.writable.getWriter()` is required for the serial port to be closed later.

```js
const writer = port.writable.getWriter();

const data = new Uint8Array([104, 101, 108, 108, 111]); // hello
await writer.write(data);

// Allow the serial port to be closed later.
writer.releaseLock();
```

Send text to the device through a `TextEncoderStream` piped to `port.writable`
as shown below.

```js
const textEncoder = new TextEncoderStream();
const writableStreamClosed = textEncoder.readable.pipeTo(port.writable);

const writer = textEncoder.writable.getWriter();

await writer.write("hello");
```

### Close a serial port {: #close-port }

`port.close()` closes the serial port if its `readable` and `writable` members
are [unlocked], meaning `releaseLock()` has been called for their respective
reader and writer.

```js
await port.close();
```

However, when continuously reading data from a serial device using a loop,
`port.readable` will always be locked until it encounters an error. In this
case, calling `reader.cancel()` will force `reader.read()` to resolve

> immediately with `{ value: undefined, done: true }` and therefore allowing loop

Reviewer: ```suggestion
immediately with `{ value: undefined, done: true }` and therefore allowing the loop
```

    break;
  }
  // value is a string.
  console.log(value);
}

const textEncoder = new TextEncoderStream();
const writableStreamClosed = textEncoder.readable.pipeTo(port.writable);

reader.cancel();
await readableStreamClosed.catch(() => { /* Ignore the error */ });

writer.close();
await writableStreamClosed;

await port.close();
```

### Listen to connection and disconnection {: #connection-disconnection }

If a serial port is provided by a USB device then that device may be connected
or disconnected from the system. When the website has been granted permission to
access a serial port, it should monitor the `connect` and `disconnect` events.

```js
navigator.serial.addEventListener("connect", (event) => {
  // TODO: Automatically open event.port or warn user a port is available.
});

navigator.serial.addEventListener("disconnect", (event) => {
  // TODO: Remove |event.port| from the UI.
  // If the serial port was opened, a stream error would be observed as well.
});
```

### Handle signals {: #signals }

After establishing the serial port connection, you can explicitly query and set
signals exposed by the serial port for device detection and flow control. These

> signals are defined as boolean values. For example, some devices like Arduino

Reviewer: ```suggestion
signals are defined as boolean values. For example, some devices such as Arduino
```

```

```js
const signals = await port.getSignals();
console.log(`Clear To Send:       ${signals.clearToSend}`);
console.log(`Data Carrier Detect: ${signals.dataCarrierDetect}`);
console.log(`Data Set Ready:      ${signals.dataSetReady}`);
console.log(`Ring Indicator:      ${signals.ringIndicator}`);
```

### Transforming streams {: #transforming-streams }

When you receive data from the serial device, you won't necessarily get all of
the data at once. It may be arbitrarily chunked. For more information, see
[Streams API concepts].

To deal with this, you can use some built-in transform streams such as
`TextDecoderStream` or create your own transform stream which allows you to
parse the incoming stream and return parsed data. The transform stream sits
between the serial device and the read loop that is consuming the stream. It can
apply an arbitrary transform before the data is consumed. Think of it like an
assembly line: as a widget comes down the line, each step in the line modifies
the widget, so that by the time it gets to its final destination, it's a fully
functioning widget.

<figure class="w-figure">
  <img src="./aeroplane-factory.jpg" alt="Photo of an aeroplane factory">

>   <figcaption class="w-figcaption">World War 2 Castle Bromwich Aeroplane Factory</figcaption>

Reviewer: ```suggestion
  <figcaption class="w-figcaption">World War II Castle Bromwich Aeroplane Factory</figcaption>
```

Reviewer: I've never seen it done with anything other than Roman numerals in English.

  constructor() {
    // A container for holding stream data until a new line.
    this.chunks = "";
  }

  transform(chunk, controller) {
    // Append new chunks to existing chunks.
    this.chunks += chunk;
    // For each line breaks in chunks, send the parsed lines out.
    const lines = this.chunks.split("\r\n");
    this.chunks = lines.pop();
    lines.forEach((line) => controller.enqueue(line));
  }

  flush(controller) {
    // When the stream is closed, flush any remaining chunks out.
    controller.enqueue(this.chunks);
  }
}
```

```js/2-5
const textDecoder = new TextDecoderStream();
const readableStreamClosed = port.readable.pipeTo(textDecoder.writable);
const reader = textDecoder.readable
  .pipeThrough(new TransformStream(new LineBreakTransformer()))
  .getReader();
```

For debugging serial device communication issues, use the `tee()` method of
`port.readable` to split the streams going to or from the serial device. The two
streams created can be consumed independently and this allows you to print one
to the console for inspection.

```js
const [appReadable, devReadable] = port.readable.tee();

// TODO: Update UI with incoming data from appReadable.

> // TODO: Log incoming data in JS console for inspection from devReadable.

Reviewer: I'm thinking "TODO" should be removed. People might think we neglected to finish the code example.

Author: Fixed.

```

## Codelab {: #codelab }

In the [Google Developer codelab], you'll use the Serial API to interact with a
[BBC micro:bit] board to show images on its 5x5 LED matrix.

## Polyfill {: #polyfill }

On Android, support for USB-based serial ports is possible using the WebUSB API
and the [Serial API polyfill]. This polyfill is limited to hardware and
platforms where the device is accessible via the WebUSB API because it has not
been claimed by a built-in device driver.

## Security and privacy {: #security-privacy }

> The Chrome team has designed and implemented the Serial API using the core

Reviewer: ```suggestion
The spec authors designed and implemented the Serial API using the core
```

Author: Fixed. But what is the rationale?

Reviewer: It seems that Reilly listed spec authors in a way that credits everyone who contributed to the spec repo in some way. Not all of them are Google employees.
